src/libs/ApiUtils.js

```diff
@@ -0,0 +1,80 @@
+import lodashGet from 'lodash/get';
+import Onyx from 'react-native-onyx';
+import ONYXKEYS from '../ONYXKEYS';
+import CONFIG from '../CONFIG';
+import CONST from '../CONST';
+import * as Environment from './Environment/Environment';
+
+// To avoid rebuilding native apps, native apps use production config for both staging and prod
+// We use the async environment check because it works on all platforms
+let ENV_NAME = CONST.ENVIRONMENT.PRODUCTION;
+let shouldUseStagingServer = false;
+Environment.getEnvironment()
+    .then((envName) => {
+        ENV_NAME = envName;
+
+        // We connect here, so we have the updated ENV_NAME when Onyx callback runs
+        Onyx.connect({
+            key: ONYXKEYS.USER,
+            callback: (val) => {
+                // Toggling between APIs is not allowed on staging
+                if (ENV_NAME === CONST.ENVIRONMENT.PRODUCTION) {
+                    shouldUseStagingServer = false;
+                    return;
+                }
+
+                const defaultToggleState = ENV_NAME === CONST.ENVIRONMENT.STAGING;
+                shouldUseStagingServer = lodashGet(val, 'shouldUseStagingServer', defaultToggleState);
+            },
+        });
+    });
```

Comment:
Updated the code and moved all the `shouldUseStaging` logic here

A follow-up PR can move this logic as a default Onyx value for the key `user. shouldUseStagingServer` so we don't have any logic here at all